paddle/fluid/operators/mkldnn/dequantize_mkldnn_op.cc

```diff
@@ -36,100 +36,58 @@ template <typename T>
-    MKLDNNMemoryFormat src_fmt = input->format();
-
-    std::string key =
-        platform::CreateKey(dev_ctx, src_dt, src_tz, ctx.OutputName("Output"));
-    key = platform::ExtendKeyWithThreadInfoIfNeeded(dev_ctx, key);
-
-    const std::string key_prim = key + "@r";
-    const std::string key_src_mem = key + "@s";
-    const std::string key_dst_mem = key + "@d";
-
-    std::shared_ptr<dnnl::memory> src_memory;
-    std::shared_ptr<dnnl::memory> dst_memory;
-    std::shared_ptr<reorder> reorder_p;
-    reorder_p = std::static_pointer_cast<reorder>(dev_ctx.GetBlob(key_prim));
-
-    if (reorder_p == nullptr) {
-      dnnl::primitive_attr attri;
-      int mask = 0;
-      float reorder_scale = 1. / scale_data;
-      attri.set_output_scales(mask, {reorder_scale});
-
-      if (with_shift) {
-        dnnl::post_ops post_operations;
-        post_operations.append_sum();
-        attri.set_post_ops(post_operations);
-        std::fill(output_data, output_data + output->numel(), reorder_shift);
-      }
-
-      auto src_md = platform::MKLDNNMemDesc({src_tz}, src_dt, src_fmt);
-      src_memory = std::make_shared<dnnl::memory>(src_md, engine,
-                                                  to_void_cast<T>(input_data));
-
-      auto dst_md =
-          platform::MKLDNNMemDesc({dst_tz}, memory::data_type::f32,
-                                  platform::MKLDNNFormatForSize(
-                                      dst_tz.size(), MKLDNNMemoryFormat::nchw));
-
-      dst_memory = std::make_shared<dnnl::memory>(
-          dst_md, engine, to_void_cast<float>(output_data));
-
-      auto reorder_pd = std::shared_ptr<reorder::primitive_desc>(
-          new reorder::primitive_desc(*src_memory, *dst_memory, attri));
-      reorder_p = std::shared_ptr<reorder>(new reorder(*reorder_pd));
-      dev_ctx.SetBlob(key_prim, reorder_p);
-      dev_ctx.SetBlob(key_src_mem, src_memory);
-      dev_ctx.SetBlob(key_dst_mem, dst_memory);
-    } else {
-      src_memory =
-          std::static_pointer_cast<dnnl::memory>(dev_ctx.GetBlob(key_src_mem));
-      src_memory->set_data_handle(to_void_cast<T>(input_data));
-
-      dst_memory =
-          std::static_pointer_cast<dnnl::memory>(dev_ctx.GetBlob(key_dst_mem));
-      if (with_shift)
-        std::fill(output_data, output_data + output->numel(), reorder_shift);
-      dst_memory->set_data_handle(output->mutable_data<float>(ctx.GetPlace()));
+
+    auto x_tz = phi::vectorize<int64_t>(x->dims());
+    auto x_paddle_dtype = framework::TransToProtoVarType(x->dtype());
+    auto out_paddle_dtype = framework::TransToProtoVarType(out->dtype());
+
+    dnnl::primitive_attr attrs;
+    static constexpr int32_t mask = 0;  // same shift and scale for whole tensor
+
+    const float reorder_scale = 1. / quantization_scale;
+    attrs.set_output_scales(mask, {reorder_scale});
+
+    if (with_shift) {
+      attrs.set_zero_points(DNNL_ARG_SRC, mask,
+                            {static_cast<int32_t>(quantization_shift)});
```

Comment:
Have you check it if zero_point works with reorder primitive. On this page https://oneapi-src.github.io/oneDNN/dev_guide_reorder.html#doxid-dev-guide-reorder  I can see that only output_scale and sum post-op is available.

Reply:
@wozna I think that there is an issue with their documentation since DNNL_VERBOSE=1 shows following when I am running quantize tests:
<pre>onednn_verbose,exec,cpu,reorder,jit:uni,undef,src_f32::blocked:ab:f0 dst_u8::blocked:ab:f0,attr-oscale:0:127.000000 <b>attr-zero-points</b>:dst:0:128 ,,2x3,0.000976562</pre>
so it seems like zero points are supported and the results of this tests are correct

Reply:
If it works that's fine. I am wondering now about shift setting in dequantize. Because if we want to perform the equation in quantize `INT8 = output_scale * (FP32 + shift)`, everything has its order, first is zero_point and then output_scale. But in dequantize we have `FP32 = INT8 / output_scale - shift`. If we follow the order, zero_point first, then output_scale, the equation looks like this `FP32 = (1 / output_scale) * (INT8 - shift * output_scale)`. So I think shift should be `quantization_shift = - scale_shift * scale_data`. Please correct me if I am wrong.
Moreover, I think that we don't have any model that uses shift in dequantize, as all RNN int8 ops have force_fp32_output, therefore it could be never been tested in the production.

Reply:
In unit test I can see that the order is `scale` first, then `zero_point`. It is important to confirm this.

Reply:
Now everything is clear.

Correct equations: 
1) **Quantize** Using reorder primitive from FP32 to INT8 with set_output_scales and set_zero_points (DST) gives following equation:
`DST = (SRC * output_scales) + zero_points` 

2) **Dequantize** Using reorder primitive from INT8 to FP32 with set_output_scales and set_zero_points (SRC) gives following equation:
`DST = (SRC - zero_points) * output_scales`